### PR TITLE
Bug 1375350 Edits to OOR info

### PR DESCRIPTION
**Pull request by bfallonf**
As per: https://bugzilla.redhat.com/show_bug.cgi?id=1375350

But also, I went through all of #2690 because I wanted to get to some of the vague bits.

@derekwaynecarr  I'll put some comments in the PR. Can I get your thoughts? And of course if you have any comments on my changes.

Thanks!

cc: @ahardin-rh

### CODE REVIEW

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: I do think this could be changed to something like "How the node signals that it's almost full", (though that's a terrible suggestion). But really, the paragraph below seems to be describing a setting that says that. Any other suggestions?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: "The node can support the ability"

Does this mean it is not enabled by default? Is this what we're enabling here?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: Where are these signals actually given out? Is this in the logs? Which files exactly?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: I think what's missing is an explanation of what an eviction threshold, and why you would want to use a soft one over a hard one.

Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: With the below, and this is said a few times, but we're saying "The node supports the ability to X" a lot. I take that to mean something needs to be enabled in order to use it. Is that the case? If it's by default, it should just be saying "The node can X".

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: Where is this found? A file somewhere?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: What does that mean? Is that the only valid value for the <operator> bit?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: How do you find that? Quantity of what?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: With the below, I think an example would be a lot better... @derekwaynecarr Do you have one we could put into the docs?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: Is cAdvisor something supported by OpenShift? I've not heard of it before. Is there a better place we could link out to?

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: I'm not sure what this is saying. Is it saying that the scheduler can read the eviction signal and do something accordingly? Is so, I'd move this above to the signal section.

File: admin_guide/overcommit.adoc
Comment (bfallonf) [inline admin_guide/overcommit.adoc]: @derekwaynecarr This seems odd. It's as though having it enabled is a bad thing. Is there a reason why it's not just disabled by default instead of the user needing to do this??

**Comment by bfallonf** [comment]
@derekwaynecarr ^ bump

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: an example for a soft eviction and hard eviction threshold would be the following:

* operator wants to evict immediately if memory falls below <5%  (hard threshold)
* operator wants to evict if memory falls below <30% for 1 min (soft threshold)

in this scenario, the operator would like their machines to steady at 70% utilization, but are willing to go above that for short periods of time.  that is the general idea.  our ops team will typically use two thresholds for a similar reason.

Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: we have no default eviction thresholds enabled in 3.3, so a user today must opt-in to this behavior.  Meaning they need to setup the node to do this now.  we will get defaults in the future.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: this is basically describing a syntax for a threshold.

in 3.3, we just support literal thresholds (memory.available<100Mi)

in 3.4, we will support percentage thresholds as well (memory.available<10%)

so in 3.4, we will need to update this doc to reflect that both input styles are valid.

a sample node configuration is given in the "Example scenario" section.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: its meant to mean the signals in table 1 (which now lists one), but in 3.4 needs to list 4 more (all disk related)

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: right now, < is the only supported operator, but there was discussion on potentially offering more.  i kept the doc this way to allow it to grow without major re-writes.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: this basically means 'the syntax you used to express a quantity anywhere in openshift/kube' - so whether its how you declare the amount of cpu or memory for a pod, a constraint on quota, a limit in a limit range, etc.  they all use the quantity representation.  maybe its just obvious.  we don't appear to have a good doc to describe: 
https://github.com/kubernetes/kubernetes/blob/master/docs/design/resources.md#resource-quantities

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: can we hold on a soft eviction scenario until we have disk?  i think that will make more sense in that context for 3.4.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: As noted earlier, I think we should just say:

"The node evaluates eviction thresholds every 10s."

In the future, hard eviction thresholds for memory will not use polling every 10s, and instead we will have the kernel tell us the threshold has been passed and act immediately.  That is planned for Kubernetes 1.5 / Origin 1.5.

Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: so drop any mention of housekeeping interval.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: just drop this since its not an actual flag that can be modified at this time.  sorry for the error on my original doc.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: this is important.   the default eviction-pressure-transition-period is 5m0s, so if you see a node that reports MemoryPressure, it will wait 5m0s to actually report that MemoryPressure is not longer occuring.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: i like that you used 'failed' consistently above, we should do the same here and s/killed/failed

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: The if is confusing here.

A `Guaranteed` pod is guaranteed to never be evicted as a result of another pod's resource consumption.

The next section is trying to say:

The only reason a `Guaranteed` pod could be evicted is as follows:

* a system daemon...
* ...

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: i think it should be "and the node only has Guaranteed pods remaining" in the same sentence.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: basically, I am trying to explain that the only rationale thing to do when a node is running only guaranteed pods but system services are consuming too much resource is to fail the guaranteed pods since i cant really fail node system services.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: I would not move this to the signal section.

The list of reported Node conditions will grow in 3.4 to include DiskPressure.

This table is saying the following:

The scheduler looks at the NodeConditions reported by the node, and if it sees the node reporting "MemoryPressure" it will not place BestEffort pods on that node.

In 3.4, if the scheduler sees nodes that report "DiskPressure", it will not schedule any pods to that node.

The list of pressure conditions will grow, and the scheduler will do something slightly different for each.

Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: so the scheduler does NOT read eviction signals, it reads Node Conditions that are driven based on the configured eviction thresholds.  So for example, if i set an eviction threshold like the following:

eviction-hard is "memory.available<500Mi"

if available memory falls below that value, the Node has a value reported in `Node.Status.Conditions[]` whose `Type` will be `MemoryPressure` and whose `Status` will be `True`.  It's that value on the node object that the scheduler integrates with when making scheduling decisions.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: I would not nest this below Troubleshooting.  I would keep it after Eviction Thresholds in a dedicated section.

File: admin_guide/out_of_resource_handling.adoc
Comment (derekwaynecarr) [inline admin_guide/out_of_resource_handling.adoc]: s/it/it's

I think this is so important that it should be called out maybe in the top of the document with something like ensuring your node has been configured correctly.

File: admin_guide/overcommit.adoc
Comment (derekwaynecarr) [inline admin_guide/overcommit.adoc]: It is a bad thing, but some customers want to have it enabled because they used that feature to meet certain densities in v2.  unfortunately, swap being enabled means you can use other features.

@sdodson -- would it be bad for us to disable swap by default in our install, and instead write doc to discuss how it could be turned back on if desired?  is that something we can look to do in 3.4/3.5?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
Thanks for improving the documentation, please address the comments.

File: admin_guide/overcommit.adoc
Comment (sdodson) [inline admin_guide/overcommit.adoc]: Yeah, it's not hard, just consensus building. We'll try to get it in the first update after 3.4. https://github.com/openshift/openshift-docs/pull/3023

Comment (bfallonf) [inline admin_guide/overcommit.adoc]: Sounds good. The initial BZ is about this contradiction, so I'll check if that's enough for Eric ( @TheDiemer ) and continue with the rest of the comments. Thanks, all.

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: Changed to "the node can be configured to..." cos 3.3 is the current.

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: OK. I get you. I made a link out to the docs above, but you're right, it might be a good idea for us on docs to doc this at some point.

File: admin_guide/out_of_resource_handling.adoc
Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: I did some minor rewrites so that it's specified that 10 seconds is the housekeeping interval.

Comment (bfallonf) [inline admin_guide/out_of_resource_handling.adoc]: Scratch that, after your next couple comments I scrapped it all instead.

**Comment by bfallonf** [comment]
@derekwaynecarr Thanks for taking a look. I've made edits, pretty much to what you suggested. Can I get a final ack there's nothing else before I move forward with this? Thanks, again.

**Comment by bfallonf** [comment]
@derekwaynecarr ^ Bump. (Thanks!)

**Comment by derekwaynecarr** [comment]
@bfallonf -- this is a big improvement.  LGTM

**Comment by bfallonf** [comment]
Big thanks @derekwaynecarr ! Good to see it's an improvement. Can I ask you to please approve the changes? Seems that new GitHub review feature means people need to give the thumbs up by clicking a button.

@adellape @ahardin-rh Any comments before I merge?

**Review by derekwaynecarr — APPROVED** [review]

**Comment by derekwaynecarr** [comment]
@bfallonf -- approved changes, if we can merge this today I can send my updates for the disk-eviction support in 1.4 origin release.

**Comment by bfallonf** [comment]
@derekwaynecarr Sure thing. Thanks much. I'll get this merged. If this has been reviewed before tomorrow morning, I can maybe get someone in BNE to take a look.

cc @adellape @ahardin-rh

File: admin_guide/out_of_resource_handling.adoc
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: With the updated doc guidelines, we can apply the new formatting here:

    `PodPhase`

File: admin_guide/out_of_resource_handling.adoc
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: Apply updated formatting here:

    `Node.Status.Conditions`
    `MemoryPressure`

File: admin_guide/out_of_resource_handling.adoc
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: `MemoryPressure`

File: admin_guide/out_of_resource_handling.adoc
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: `oom_score_adj`

File: admin_guide/out_of_resource_handling.adoc
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: `oom_score_adj`

File: admin_guide/out_of_resource_handling.adoc
Comment (ahardin-rh) [inline admin_guide/out_of_resource_handling.adoc]: `BestEffort`

**Comment by ahardin-rh** [comment]
@bfallonf just a few minor comments from me regarding updated style guidelines. :star:

**Comment by bfallonf** [comment]
Thanks @ahardin-rh . I'll pay more attention to the new guidelines... I'll merge away!

**Pull request merged**
